### PR TITLE
Resources

### PR DESCRIPTION
**Pull request by bergman**
Missing:
 - [ ] cli
 - [x] tests

### CODE REVIEW

File: styx-api-service/src/main/java/com/spotify/styx/api/CliResource.java [hunk 1]
```diff
@@ -96,11 +94,21 @@ public CliResource(String schedulerServiceBaseUrl, EventStorage eventStorage) {
...
+            rc -> proxy("/" + arg("endpoint", rc), rc)),
+        Route.async(
+            "GET", BASE + "/<endpoint>/<id>",
+            rc -> proxy("/" + arg("endpoint", rc) + "/" + arg("id", rc), rc)),
+        Route.async(
+            "DELETE", BASE + "/<endpoint>/<id>",
+            rc -> proxy("/" + arg("endpoint", rc) + "/" + arg("id", rc), rc)),
+        Route.async(
+            "PATCH", BASE + "/<endpoint>/<id>",
+            rc -> proxy("/" + arg("endpoint", rc) + "/" + arg("id", rc), rc))
```
Comment (rouzwawi) [inline styx-api-service/src/main/java/com/spotify/styx/api/CliResource.java]: You might want to replace these all with just a `/<endpoint:path>` parameter that will match the rest of the uri. see https://github.com/spotify/apollo/blob/master/apollo-api/docs/routing-engine.md#path-parameters

File: styx-api-service/src/main/java/com/spotify/styx/api/ResourceResource.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    if (!payload.isPresent()) {
+      return Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Missing payload."));
+    }
+
+    Resource resource;
+    try {
+      resource = OBJECT_MAPPER.readValue(payload.get().toByteArray(), Resource.class);
+    } catch (IOException e) {
+      return Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Invalid payload."));
+    }
```
Comment (rouzwawi) [inline styx-api-service/src/main/java/com/spotify/styx/api/ResourceResource.java]: Use the EntityMiddleware to avoid having to do manual parsing and error handling.

File: styx-common/src/main/java/com/spotify/styx/model/DataEndpoint.java [hunk 3]
```diff
@@ -49,15 +49,19 @@
...
+
   @JsonCreator
   public static DataEndpoint create(
       @JsonProperty("id") String id,
       @JsonProperty("partitioning") Partitioning partitioning,
       @JsonProperty("docker_image") Optional<String> dockerImage,
       @JsonProperty("docker_args") Optional<List<String>> dockerArgs,
-      @JsonProperty("secret") Optional<Secret> secret) {
+      @JsonProperty("secret") Optional<Secret> secret,
+      @JsonProperty("resources") Optional<List<String>> resources) {
```
Comment (rouzwawi) [inline styx-common/src/main/java/com/spotify/styx/model/DataEndpoint.java]: I think jackson can handle parsing a missing key directly into an empty list, without having to wrap it in an optional. For the `resources` field, a missing key and an empty list should have the same meaning.

File: styx-common/src/main/java/com/spotify/styx/model/Resource.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ */
+@AutoValue
+@JsonIgnoreProperties(ignoreUnknown = true)
+public abstract class Resource {
+
+  @JsonProperty
+  public abstract String id();
+
+  @JsonProperty
+  public abstract long concurrency();
```
Comment (rouzwawi) [inline styx-common/src/main/java/com/spotify/styx/model/Resource.java]: Do we want to call it `concurrency` or `limit`? "Resource limit" sounds more natural to me than a "Resource concurrency".

File: styx-api-service/src/main/java/com/spotify/styx/api/ResourceResource.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            json(), "POST", BASE,
+            rc -> postResource(rc.request())),
+        Route.with(
+            json(), "GET", BASE + "/<rid>",
+            rc -> getResource(arg("rid", rc))),
+        Route.with(
+            json(), "DELETE", BASE + "/<rid>",
+            rc -> deleteResource(arg("rid", rc))),
+        Route.with(
+            json(), "PATCH", BASE + "/<rid>",
```
Comment (rouzwawi) [inline styx-api-service/src/main/java/com/spotify/styx/api/ResourceResource.java]: Shouldn't this be PUT?

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java [hunk 6]
```diff
@@ -451,7 +451,8 @@ private TriggerListener trigger(
 
     return (workflow, triggerId, instant) -> {
       try {
-        if (!storage.enabled(workflow.id()) || !storage.globalEnabled()) {
+        final Set<WorkflowId> enabledWorkflows = storage.enabledWorkflows();
```
Comment (rouzwawi) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: Unrelated change? Why get all the enabled workflows instead of just checking for a specific one?

Comment (rouzwawi) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: If it was changed because `enabled(WorkflowId)` is deprecated, then use `storage.workflowState` instead.

File: styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java [hunk 7]
```diff
@@ -477,4 +482,39 @@ private static DateTime instantToDatetime(Instant instant) {
   private static Instant datetimeToInstant(DateTime dateTime) {
     return dateTime.toDate().toInstant();
   }
+
+  Resource getResource(String id) {
+    return entityToResource(
+        datastore.get(datastore.newKeyFactory().kind(KIND_RESOURCE).newKey(id)));
```
Comment (rouzwawi) [inline styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java]: Potential NPE if the lookup does not find a result.

File: styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java [hunk 8]
```diff
@@ -477,4 +482,39 @@ private static DateTime instantToDatetime(Instant instant) {
...
+  }
+
+  void postResource(Resource resource) {
+    datastore.put(resourceToEntity(resource));
```
Comment (rouzwawi) [inline styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java]: Wrap in retries?

File: styx-common/src/main/java/com/spotify/styx/storage/InMemStorage.java [hunk 9]
```diff
@@ -20,9 +20,11 @@
 
 package com.spotify.styx.storage;
 
+import com.google.bigtable.repackaged.com.google.common.collect.ImmutableList;
```
Comment (rouzwawi) [inline styx-common/src/main/java/com/spotify/styx/storage/InMemStorage.java]: Shaded import

File: styx-common/src/main/java/com/spotify/styx/storage/Storage.java [hunk 10]
```diff
@@ -116,14 +117,14 @@
    * @return true if the queried workflow is enabled
    */
   @Deprecated
-  boolean enabled(WorkflowId workflowId) throws IOException;
+  boolean isWorkflowEnabled(WorkflowId workflowId) throws IOException;
```
Comment (rouzwawi) [inline styx-common/src/main/java/com/spotify/styx/storage/Storage.java]: These were named with the argument type in mind, like `store(Workflow)`, `delete(WorkflowId)`, `enabled(WorkflowId)`.

The names are all repetitive now with `storeWorkflow(Workflow)`.

I can see that the last one is be better as `isEnabled(WorkflowId)`, though.

If we want to rename them, then maybe the metric names in `MeteredStorage` should be aligned too.

File: styx-api-service/src/main/java/com/spotify/styx/api/CliResource.java (hunk 1 above)
Comment (bergman) [inline styx-api-service/src/main/java/com/spotify/styx/api/CliResource.java]: Thanks, was thinking of that but didn't know how to do it :)

File: styx-common/src/main/java/com/spotify/styx/model/Resource.java (hunk 4 above)
Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/model/Resource.java]: The word `limit` might not convey that it gets replenished after use. Maybe we can look into what wording thread pools use?

File: styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java (hunk 8 above)
Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java]: Thanks!

File: styx-common/src/main/java/com/spotify/styx/storage/InMemStorage.java (hunk 9 above)
Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/storage/InMemStorage.java]: Thanks!

File: styx-common/src/main/java/com/spotify/styx/storage/Storage.java (hunk 10 above)
Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/storage/Storage.java]: Ah, forgot about the strings in `MeteredStorage`. Will have another pass at this.

Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/storage/Storage.java]: Will keep `enabledWorkflows` and `isEnabled` but revert the others

Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/storage/Storage.java]: Will revert all names :)

File: styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java [hunk 11]
```diff
@@ -477,4 +482,43 @@ private static DateTime instantToDatetime(Instant instant) {
...
+
+  Optional<Resource> getResource(String id) {
+    Entity entity = datastore.get(datastore.newKeyFactory().kind(KIND_RESOURCE).newKey(id));
+    if (entity == null) {
+      return Optional.empty();
+    }
+    return Optional.of(entityToResource(entity));
+  }
+
+  void postResource(Resource resource) throws IOException {
```
Comment (rouzwawi) [inline styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java]: No transaction needed when only doing one operation

Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java]: Fixed, thanks!

**Comment by rouzwawi** [comment]
@bergman should we wait for cli and tests or are you doing them as separate prs?

**Comment by bergman** [comment]
Let's do CLI and tests in a separate PR if you need this now. Just know that this code has never been actually run against datastore (mocked or otherwise).

**Comment by rouzwawi** [comment]
Then we should at least do tests in this PR.

File: styx-api-service/src/test/java/com/spotify/styx/api/ResourceResourceTest.java [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+      throws InterruptedException, java.util.concurrent.ExecutionException,
+             java.util.concurrent.TimeoutException {
+    return completionStage.toCompletableFuture().get(5, TimeUnit.SECONDS);
+  }
+
+  private String path(String path) {
+    return Api.Version.V1.prefix() + ResourceResource.BASE + path;
+  }
+
+  private Connection setupBigTableMockTable() {
```
Comment (rouzwawi) [inline styx-api-service/src/test/java/com/spotify/styx/api/ResourceResourceTest.java]: Do we have to set up mocked behaviour for bigtable for this test? I think it would be enough just to send in a plain mock, since none of the paths will interact with bigtable.

File: styx-api-service/src/test/java/com/spotify/styx/api/ResourceResourceTest.java [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  @Test
+  public void shouldListMultipleResources() throws Exception {
+    storage.storeResource(RESOURCE_2);
+    Response<ByteString> response =
+        awaitResponse(serviceHelper.request("GET", path("")));
+
+    assertThat(response, hasStatus(belongsToFamily(StatusType.Family.SUCCESSFUL)));
+    System.out.println("response.payload().get().utf8() = " + response.payload().get().utf8());
```
Comment (rouzwawi) [inline styx-api-service/src/test/java/com/spotify/styx/api/ResourceResourceTest.java]: println

File: styx-api-service/src/test/java/com/spotify/styx/api/ResourceResourceTest.java [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+        ImmutableMap.of("id", "resource1", "concurrency", 1),
+        ImmutableMap.of("id", "resource2", "concurrency", 2)
+    ));
+
+    // change resource2
+    Response<ByteString> putResponse =
+        awaitResponse(serviceHelper.request("PUT", path("/resource2"),
+            ByteString.encodeUtf8("{\"id\": \"resource2\", \"concurrency\": 3}")));
+    assertThat(putResponse, hasStatus(belongsToFamily(StatusType.Family.SUCCESSFUL)));
+    assertJson(putResponse, "id", equalTo("resource2"));
```
Comment (rouzwawi) [inline styx-api-service/src/test/java/com/spotify/styx/api/ResourceResourceTest.java]: Try not to assert things that are not essential for the test. Checking the put response object is already done in the `shouldUpdateResource` test case.

Comment (bergman) [inline styx-api-service/src/test/java/com/spotify/styx/api/ResourceResourceTest.java]: yup, same for first listing i guess, it's the same case as `shouldListMultipleResources` except that one doesn't `POST` but uses datastore directly

Comment (rouzwawi) [inline styx-api-service/src/test/java/com/spotify/styx/api/ResourceResourceTest.java]: Yeah I was thinking about that but then thought that it makes sense to test that the changes are actually reflected in the api. But checking the put response is not directly related.

**Comment by rouzwawi** [comment]
LGTM 👍  :shipit:
